iceoryx_examples/iceperf/iceperf_leader.cpp

```diff
@@ -0,0 +1,151 @@
+
+#include "iceperf_leader.hpp"
+#include "iceoryx.hpp"
+#include "iceoryx_c.hpp"
+#include "iceoryx_posh/popo/publisher.hpp"
+#include "iceoryx_posh/popo/subscriber.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+#include "iceoryx_utils/cxx/convert.hpp"
+#include "mq.hpp"
+#include "topic_data.hpp"
+#include "uds.hpp"
+
+#include <iomanip>
+#include <iostream>
+
+//! [use constants instead of magic values]
+constexpr const char APP_NAME[]{"iceperf-bench-leader"};
+constexpr const char PUBLISHER[]{"Hardy"};
+constexpr const char SUBSCRIBER[]{"Laurel"};
+//! [use constants instead of magic values]
+
+IcePerfLeader::IcePerfLeader(const PerfSettings settings) noexcept
+    : m_settings(settings)
+{
+    //! [cleanup outdated resources]
+#ifndef __APPLE__
+    MQ::cleanupOutdatedResources(PUBLISHER, SUBSCRIBER);
+#endif
+    UDS::cleanupOutdatedResources(PUBLISHER, SUBSCRIBER);
+    //! [cleanup outdated resources]
+}
+
+//! [do the measurement for a single technology]
+void IcePerfLeader::doMeasurement(IcePerfBase& ipcTechnology) noexcept
+{
+    ipcTechnology.initLeader();
+
+    std::vector<double> latencyInMicroSeconds;
+    const std::vector<uint32_t> payloadSizesInKB{1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096};
+    std::cout << "Measurement for:";
+    const char* separator = " ";
+    for (const auto payloadSizeInKB : payloadSizesInKB)
+    {
+        std::cout << separator << payloadSizeInKB << " kB" << std::flush;
+        separator = ", ";
+        auto payloadSizeInBytes = payloadSizeInKB * IcePerfBase::ONE_KILOBYTE;
+
+        ipcTechnology.preLatencyPerfTestLeader(payloadSizeInBytes);
+
+        auto latency = ipcTechnology.latencyPerfTestLeader(m_settings.numberOfSamples);
+
+        latencyInMicroSeconds.push_back(latency);
+
+        ipcTechnology.postLatencyPerfTestLeader();
+    }
+    std::cout << std::endl;
+
+    ipcTechnology.releaseFollower();
+
+    ipcTechnology.shutdown();
+
+    std::cout << std::endl;
+    std::cout << "#### Measurement Result ####" << std::endl;
+    std::cout << m_settings.numberOfSamples << " round trips for each payload." << std::endl;
+    std::cout << std::endl;
+    std::cout << "| Payload Size [kB] | Average Latency [µs] |" << std::endl;
+    std::cout << "|------------------:|---------------------:|" << std::endl;
+    for (size_t i = 0U; i < latencyInMicroSeconds.size(); ++i)
+    {
+        std::cout << "| " << std::setw(17) << payloadSizesInKB.at(i) << " | " << std::setw(20) << std::setprecision(2)
+                  << latencyInMicroSeconds.at(i) << " |" << std::endl;
+    }
+
+    std::cout << std::endl;
+    std::cout << "Finished!" << std::endl;
+}
+//! [do the measurement for a single technology]
+
+//! [run all technologies]
+int IcePerfLeader::run() noexcept
+{
+    iox::runtime::PoshRuntime::initRuntime(APP_NAME);
+
+    //! [send setting to follower application]
+    iox::capro::ServiceDescription serviceDescription{"IcePerf", "Settings", "Comedians"};
```

Comment:
Maybe we should rename the event to `latency` and `throughput` so that it is transparent on consistent on all levels (see introspection).